core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerContainer.scala

```diff
@@ -66,27 +59,25 @@ object DockerContainer {
              environment: Map[String, String] = Map(),
              network: String = "bridge",
              dnsServers: Seq[String] = Seq(),
-             name: Option[String] = None)(implicit docker: DockerApiWithFileAccess,
-                                          runc: RuncApi,
-                                          ec: ExecutionContext,
-                                          log: Logging): Future[DockerContainer] = {
+             name: Option[String] = None,
+             dockerRunParameters: MultiMap[String, String])(implicit docker: DockerApiWithFileAccess,
+                                                            runc: RuncApi,
+                                                            ec: ExecutionContext,
+                                                            log: Logging): Future[DockerContainer] = {
     implicit val tid = transid
 
-    val environmentArgs = environment.map {
+    val environmentArgs = environment.flatMap {
       case (key, value) => Seq("-e", s"$key=$value")
-    }.flatten
+    }
 
-    val dnsArgs = dnsServers.map(Seq("--dns", _)).flatten
+    val params = dockerRunParameters.flatMap {
+      case (key, valueList) =>
+        valueList.map {
```

Comment:
Could be `flatMap` to not iterate again.

Reply:
FlatMap on valueList (a Set) caused the duplicate keys to be removed - I changed it to `valueList.toList.flatMap` to avoid that, but I'm not sure that is what you're looking for, performance-wise. Let me know if you have suggestions.